Packet++/src/NtpLayer.cpp

```diff
@@ -0,0 +1,516 @@
+
+        // Offset change should be done here because of overflow
+        return integerPart + fractionPart - EPOCH_OFFSET;
+    }
+
+    uint32_t NtpLayer::convertToShortFormat(const double val)
+    {
+        uint32_t retval = 0;
+        double integerPart, fractionPart;
+        uint16_t integerPartInt, fractionPartInt;
+
+        fractionPart = modf(val, &integerPart);
+
+        // Cast values to 16bit
+        integerPartInt = hostToNet16(integerPart);
+        fractionPartInt = hostToNet16(fractionPart * NTP_FRIC);
+
+        retval = retval | uint32_t(integerPartInt);
+        retval = retval | (uint32_t(fractionPartInt)) << 16;
+
+        return retval;
+    }
+
+    uint64_t NtpLayer::convertToTimestampFormat(const double val)
+    {
+        uint64_t retval = 0;
+        double integerPart, fractionPart;
+        uint32_t integerPartInt, fractionPartInt;
+
+        fractionPart = modf(val, &integerPart);
+
+        // Cast values to 32bit
+        integerPartInt = hostToNet32(integerPart + EPOCH_OFFSET);
+        fractionPartInt = hostToNet32(fractionPart * NTP_FRAC);
+
+        retval = retval | uint64_t(integerPartInt);
+        retval = retval | (uint64_t(fractionPartInt) << 32);
+
+        return retval;
+    }
+
+    std::string NtpLayer::convertToIsoFormat(const double timestamp)
+    {
+        char buffer[50], bufferFraction[15];
+        double integerPart, fractionPart;
+        struct tm *timer;
+        time_t timeStruct;
+
+        fractionPart = modf(timestamp, &integerPart);
+
+        timeStruct = integerPart;
+#if defined(WIN32) || defined(WINx64)
+        if (timeStruct < 0)
+            timeStruct = 0;
+        timer = gmtime(&timeStruct);
+#else
+        struct tm timer_r;
+        timer = gmtime_r(&timeStruct, &timer_r);
+
+        if (timer != NULL)
+            timer = &timer_r;
+#endif
+        if(timer == NULL)
+        {
+            LOG_ERROR("Can't convert time");
+            return std::string();
+        }
+        strftime(buffer, sizeof(buffer) - sizeof(bufferFraction), "%Y-%m-%dT%H:%M:%S", timer);
+
+        snprintf(bufferFraction, sizeof(bufferFraction), "%.09fZ", abs(fractionPart));
+        strncat(buffer, &bufferFraction[1], sizeof(bufferFraction));
+
+        return std::string(buffer);
+    }
+
+    std::string NtpLayer::convertToIsoFormat(const uint64_t timestampInNTPformat)
+    {
+        return convertToIsoFormat(convertFromTimestampFormat(timestampInNTPformat));
+    }
+
+    bool NtpLayer::isDataValid(const uint8_t *data, size_t dataSize)
+    {
+        if (data && dataSize >= sizeof(ntp_header))
+            return true;
+        return false;
+    }
+
+    std::string NtpLayer::toString() const
+    {
+        std::stringstream ss;
+
+        ss << "NTP Layer, ";
+        ss << "Version: " << (int)getVersion() << ", Mode: " << (int)getMode() << ", ";
+        ss << "Leap Indicator: " << (int)getLeapIndicator() << ", Stratum: " << (int)getStratum();
```

Comment:
maybe to keep this string short we can do something similar to Wireshark:
`NTP Layer v4, mode: client`

Reply:
Right now it returns an empty string:

```cpp
std::string NtpLayer::toString() const
{
    return std::string();
    //std::stringstream ss;

    //ss << "NTP Layer v" << (int)getVersion() << ", Mode: " << getModeString();

    //return ss.str();
}
```

I guess we should fix it?